Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1378 @@
+  * `DeriveFoldable` allows us to write `deriving (Foldable)`, which will be helpful later.
+  * We also import some extra things that will be useful in utilities later.
+* The following lines instruct Happy on how exactly to generate the parser. The things we should provide to it include:
+  * One or more names for our parser. The syntax should be `%name PARSER_NAME [PRODUCTION]`. If the `PRODUCTION` is not given, it uses the first non-terminal that appears in the file.
+  * What is the type of our tokens with `%tokentype`.
+  * Which function should it call if there is a parse error with `%error`.
+  * What is the monad that should be used and its canonical operations with `%monad` (we reuse `Alex` for simplicity).
+  * Which function lexes tokens and how to extract the EOF token with `%lexer`.
+* `%%`, after which we must specify the production rules for the grammar.
+* The bottom of the file (trailer) to include more Haskell definitions, such as auxiliary files and the AST that will be built by the parser. You can also define those in another module if you prefer.
+
+### Tokens
+
+Since Happy operates on tokens, we must tell it how to match each token.
+This is done by using `%token`, which must be placed before `%%` in the file.
+For each token, on the left side, we put an alias that we'll use to refer to our terminals.
+On the right side, we put a Haskell pattern between braces that indicates how Happy should match the Alex tokens.
+
+```happy
+%token
+  -- Identifiers
+  identifier { L.RangedToken (L.Identifier _) _ }
+  -- Constants
+  string     { L.RangedToken (L.String _) _ }
+  integer    { L.RangedToken (L.Integer _) _ }
+  -- Keywords
+  let        { L.RangedToken L.Let _ }
+  in         { L.RangedToken L.In _ }
+  if         { L.RangedToken L.If _ }
+  then       { L.RangedToken L.Then _ }
+  else       { L.RangedToken L.Else _ }
+  -- Arithmetic operators
+  '+'        { L.RangedToken L.Plus _ }
+  '-'        { L.RangedToken L.Minus _ }
+  '*'        { L.RangedToken L.Times _ }
+  '/'        { L.RangedToken L.Divide _ }
+  -- Comparison operators
+  '='        { L.RangedToken L.Eq _ }
+  '<>'       { L.RangedToken L.Neq _ }
+  '<'        { L.RangedToken L.Lt _ }
+  '<='       { L.RangedToken L.Le _ }
+  '>'        { L.RangedToken L.Gt _ }
+  '>='       { L.RangedToken L.Ge _ }
+  -- Logical operators
+  '&'        { L.RangedToken L.And _ }
+  '|'        { L.RangedToken L.Or _ }
+  -- Parenthesis
+  '('        { L.RangedToken L.LPar _ }
+  ')'        { L.RangedToken L.RPar _ }
+  -- Lists
+  '['        { L.RangedToken L.LBrack _ }
+  ']'        { L.RangedToken L.RBrack _ }
+  ','        { L.RangedToken L.Comma _ }
+  -- Types
+  ':'        { L.RangedToken L.Colon _ }
+  '->'       { L.RangedToken L.Arrow _ }
+```
+
+These aliases to tokens are the terminals we'll use while writing the grammar. They are first-class, and we can use them among other symbols.
+
+### Semantic actions
```

Comment:
Given that there are no semantic actions here, this sub-heading might not be accurate. 

Could I suggest `### First production rule`?